[libknet] allow better handling of internal threads

fabbione opened the pull request:
while playing around adding some dynamic internal threads, i did
hit a limitation with allocating a dynamic THREAD_MAX value based
on OS availability of certain features.

this patch changes from dynamic to static allocation of threads, that
can always be increased if necessary without breaking anything.

simplify thread numbering and conditionals on OS features

improve debugging of thread status by adding a 2 phase thread registration
process:

- register a thread before starting it with THREAD_REGISTERED before pthread_create
- set a thread status as started with THREAD_STARTED

any failure in between can now be detected in debugging logs.

Note for developers: all threads *MUST* be registered/started before
wait_all_threads_status check in handle.c.

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione commented:
@wferi I am assigning those reviews to you because everybody is on xmas vacation. If you are not comfy in reviewing those patches that is perfectly fine :)

This is backportable to stable1 too.

fabbione commented:
retest this

fabbione commented:
retest this please

chrissie-c commented on libknet/threads_common.h:
```diff
@@ -14,22 +14,24 @@
-#define KNET_THREAD_STATUS_MAX	KNET_THREAD_RUNNING + 1
+#define KNET_THREAD_UNREGISTERED	0 /* we don´t know if this thread exists,
+					     we need to make sure all threads are started
+					     before we call wait_all_threas_status at init time */
```
typo in the comment

chrissie-c reviewed (dismissed):
There's a comment typo after the define for KNET_THREAD_REGISTERED, but otherwise looks good.

fabbione commented:
@chrissie-c reworked the comments a bit, hopefully they are more clear and no typos this time :)

fabbione committed:
[libknet] allow better handling of internal threads

while playing around adding some dynamic internal threads, i did
hit a limitation with allocating a dynamic THREAD_MAX value based
on OS availability of certain features.

this patch changes from dynamic to static allocation of threads, that
can always be increased if necessary without breaking anything.

simplify thread numbering and conditionals on OS features

improve debugging of thread status by adding a 2 phase thread registration
process:

- register a thread before starting it with THREAD_REGISTERED before pthread_create
- set a thread status as started with THREAD_STARTED

any failure in between can now be detected in debugging logs.

Note for developers: all threads *MUST* be registered/started before
wait_all_threads_status check in handle.c.

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

chrissie-c reviewed (approved):
ACK
